### PR TITLE
Add block indexing to Indexer API

### PR DESCRIPTION
**Pull request by aaronbuchwald**
This PR adds block indexing to the indexer API.

The new API serves both the block JSON and the `ExecutedBlock` bytes to support both https://github.com/ava-labs/hypersdk/issues/1513 and decoding the response in the golang indexer client.

We can't unmarshal the block's JSON as is because it requires unmarshalling into the `Action` and `Auth` interfaces within the transaction. This PR includes both the JSON and binary in the response to support both cases.

### CODE REVIEW

File: api/indexer/server.go [hunk 1]
```diff
@@ -44,6 +45,62 @@ func (f *apiFactory) New(vm api.VM) (api.Handler, error) {
...
+}
+
+type GetBlockByHeightRequest struct {
+	Height uint64 `json:"height"`
+}
+
+type GetBlockResponse struct {
+	Block      *chain.ExecutedBlock `json:"block"`
+	BlockBytes codec.Bytes          `json:"blockBytes"`
+}
```
Comment (aaronbuchwald) [inline api/indexer/server.go:59]: Alternative here is to include the requested encoding in the request

**Comment by containerman17** [comment]
## 1. GetBlock improvements

```js
{
    "jsonrpc": "2.0",
    "result": {
        "block": {
            "blockID": "2NmFKR118wRQc88qgmgeLqjz125UWxRb7jsv9nFNj6ygfedKyL",
            "block": {
                "parent": "qsFoFWSJkFgXgtWmh8MtyVgdTtzv77qNyjJPB7Fc43UXeUUKj",
                "timestamp": 1727846356939,
                "height": 3420,
                "txs": [
                    {
                        // 1.4 Transaction ID  
                        // 1.5 Transaction error
                        "base": {
                            "timestamp": 1727846413000,
                            "chainId": "2gdnF9MkpFFBsmFQHSpDrnXT5N3DGKXPmBJfradUyE3kzHc4QE",
                            "maxFee": 10000000
                        },
                        "actions": [
                            {// 1.2 Action names or IDs
                                "to": "0x0000000000000000000000000000000000000000000000000000000000deadc0de",
                                "value": 123456789,
                                "memo": "THVpZ2k="
                            }
                        ],
                        "auth": {
                            "signer": [ // 1.3 Auth.* should be hex
                                70,
                                220,
                                217,
                                //... more bytes here
                            ],
                            "signature": [ // 1.3 Auth.* should be hex
                                103,
                                151,
                                155,
                                //... more bytes here
                            ]
                        }
                    }
                ],
                "stateRoot": "28TpB5nFDvBYZWfEZesFpDA8zLic4MnXYwniEFDKvpkWkiBKYD"
            },
            "results": [
                { // 1.6 Result fields should be lowercase
                    "Success": true,
                    "Error": "",
                    "Outputs": [
                        "AAAAAAJMr6QTAAAAABYTZz8="
                    ],
                    // 1.1 Units and unit prices should be an array or object
                    "Units": "(Bandwidth=197, Compute=7, Storage(Read)=14, Storage(Allocate)=50, Storage(Write)=26)", 
                    "Fee": 29400
                }
            ],
            // 1.1 Units and unit prices should be an array or object
            "unitPrices": "(Bandwidth=100, Compute=100, Storage(Read)=100, Storage(Allocate)=100, Storage(Write)=100)"
        },
        "blockBytes": "000001196ef5112686d9fcfc4b3a3d2f2d439e1" //shortened
    },
    "id": 2652924454900838
}
```

## 1.1 Units and unit prices format
Should be an array or object. Binary format (see `fees.UnpackDimensions`) would work too. If needed, I can write a parser for the current string format too.

## 1.2 Action names or IDs
There is action payload, but no action name or ID. The ideal solution is to provide action binary, including the first TypeID byte, as discussed.

## 1.3 Auth.* as hex
Currently an array of numbers. It should be in hex format (note that in Golang, []byte serializes as base64 by default).

## 1.4 Transaction ID
Transaction ID is missing.

## 1.5 Transaction error
A transaction error field is missing, even when the transaction returns an error. Please use a regular string format instead of base64.

## 1.6 Result fields should be lowercase
For consistency across APIs.

**Comment by containerman17** [comment]
## 2. GetTX improvements
```js
{
    "jsonrpc": "2.0",
    "result": {
        // 2.1 Error field 
        "timestamp": 1727848107595,
        "success": false,
        "units": "(Bandwidth=197, Compute=7, Storage(Read)=14, Storage(Allocate)=50, Storage(Write)=26)", // 2.2 Units format
        "fee": 29400,
        "result": []
    },
    "id": 9669870267519104
}
```

### 2.1 Error field
Although this transaction failed, there is no error field in the JSON. As mentioned earlier, please use a regular Unicode string for the error, not base64.

### 2.2 Units and unit prices format
Same as 1.1 above, it should be formatted as an array or object.

**Comment by aaronbuchwald** [comment]
Ok here are the changes I'm planning to make as a follow on to this PR:

# Tx JSON Marshalling

- include txID in JSON
- marshal actions and auth as bytes and encode them as hex via `codec.Bytes`
- Add `TransactionJSON` intermediate type that supports `MarshalJSON` and `UnmarshalJSON` and stores the action/auth as bytes
- add `MarshalJSON` implementation to `chain.Transaction`
- add `UnmarshalJSON` that takes `chain.Parser` in as a parameter so that it can parse the actions/auth

# Results

- Add JSON annotations to lowercase the fields of `chain.Result`
- Add `ResultJSON` intermediate representation that supports json marshalling and add `MarshalJSON` and `UnmarshalJSON` to `chain.Result` that uses the intermediate type (will use `codec.Bytes` for the error result)
- Already added JSON marshalling for `fees.Dimensions` here: https://github.com/ava-labs/hypersdk/pull/1622

Side note (out of scope for this PR): we currently use an error string for the result. There's a weird edge case here where if we or a developer changes an error string that is actually part of chain processing. This should be clear, but I've found that it's easy to forget this and think we could make it more explicit. We could, for example, include errors as possible return values in the ABI.

# Block Requests

- Add `encoding` parameter to block request types
- Use a single block response type w/ `omitempty` so that it sends JSON with either the block bytes or JSON of the block

Note: golang clients can use binary representation, so that it can decode with `chain.Parser` argument and other languages can handle the JSON or binary if they choose

File: api/indexer/server.go (hunk 1 above)
Comment (containerman17) [reply api/indexer/server.go:59]: It might create a mess. There aren't that many methods. It's fine to duplicate unless we plan to increase the number of methods.

File: api/indexer/client.go [hunk 2]
```diff
@@ -27,6 +29,54 @@ type Client struct {
 	requester *requester.EndpointRequester
 }
 
+// Use a separate type that only decodes the block bytes because we cannot decode block JSON
+// due to Actions/Auth interfaces included in the block's transactions.
+type GetBlockClientResponse struct {
+	BlockBytes codec.Bytes `json:"blockBytes"`
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply api/indexer/client.go:35]: Yup

File: api/indexer/indexer.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+	"github.com/ava-labs/hypersdk/chain"
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/consts"
+	"github.com/ava-labs/hypersdk/event"
+	"github.com/ava-labs/hypersdk/fees"
+	"github.com/ava-labs/hypersdk/internal/cache"
+	"github.com/ava-labs/hypersdk/internal/pebble"
+)
+
+const maxBlockWindow uint64 = 1_000_000
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply api/indexer/indexer.go:26]: We don't maintain a blockID -> height mapping on disk to avoid using hashes as keys. This means we iterate over all the data on startup, so we set a reasonable guardrail.

If you want to store all transactions, you need to scale horizontally rather than writing to a single pebbledb instance. This won't work at scale, so I'd prefer to set a reasonable guardrail than allow users to try and scale something that wasn't meant to.

File: api/indexer/indexer.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+func (*Indexer) storeTransaction(
+	batch database.KeyValueWriter,
+	txID ids.ID,
+	timestamp int64,
+	success bool,
+	units fees.Dimensions,
+	fee uint64,
+	outputs [][]byte,
+) error {
+	outputLength := consts.ByteLen // Single byte containing number of outputs
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply api/indexer/indexer.go:194]: I agree and think storing the entire tx + result makes sense here https://github.com/ava-labs/hypersdk/pull/1626

This is not changed in this PR though, so don't think it needs to be in scope

File: api/indexer/indexer_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ava-labs/hypersdk/chain"
+	"github.com/ava-labs/hypersdk/chain/chaintest"
+	"github.com/ava-labs/hypersdk/fees"
+)
+
+func TestBlockIndex(t *testing.T) {
```
Comment (joshua-kim) [inline api/indexer/indexer_test.go:17]: This test feels like it's testing a lot of different cases which is making it complex to read/debug/add to. I feel like testing more targeted code edge-cases instead of a single god-test would make sense here.

Comment (aaronbuchwald) [reply api/indexer/indexer_test.go]: Moved a helper to generate a sequence of empty executed blocks to `chaintest` and separated out the restart portion of the test.

File: api/indexer/client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply api/indexer/client.go:35]: This is to support clients (JS) that prefer to handle JSON than the binary representation. For golang, we actually cannot handle the JSON using the current requester because it will simply unmarshal the json into the struct and we can't unmarshal JSON directly into the block type due to the `Action` and `Auth` interfaces.

Have a note here https://github.com/ava-labs/hypersdk/pull/1606#issuecomment-2389723597 to add an encoding parameter that we could block this PR on if you prefer.

**Review by joshua-kim — APPROVED** [review]
